### PR TITLE
Pagination feature/indexing waku messages

### PR DESCRIPTION
**Pull request by staheri14**
This PR addresses this issue https://github.com/status-im/nim-waku/issues/232

### CODE REVIEW

**Comment by oskarth** [comment]
@staheri14 Feel free to convert these into ready for review once you want people to have a look. Also I notice some conflicting files, probably a good idea to rebase / keep up to date with master to make merging easier!

**Comment by oskarth** [comment]
What's left to get this across the finish line @staheri14?

**Comment by staheri14** [comment]
> What's left to get this across the finish line @staheri14?
The core logic is implemented, but the followings are left:
- Checking and removing duplicate waku messages (I am not sure this fits in the scope of the current PR as in the store protocol without pagination, all the messages get stored regardless of duplication)
- Finalizing the test for the integration of pagination in the store protocol  (the test is somewhat ready just needs the approval of this PR https://github.com/status-im/nim-waku/pull/237 to be finalized)
- Investigating and testing more corner cases
- Handling history responses with no pagingInfo
- Some of the datatypes are going to change due to https://github.com/status-im/nim-waku/pull/237, which will result in some changes in the current PR as well
- Adding missing comments and documentations

**Comment by oskarth** [comment]
Other PR merged. Please rebase on master and resolve conflicts

File: tests/v2/test_waku_pagination.nim [hunk 1]
```diff
@@ -1,11 +1,21 @@
...
-  std/unittest,
+  std/[unittest,algorithm],
+  nimcrypto/sha2,
+  stew/byteutils,
   ../../waku/node/v2/waku_types,
+  ../../waku/protocol/v2/waku_store,
   ../test_helpers
 
+
+proc CreateSampleList( s : int) : seq[IndexedWakuMessage] =
```
Comment (oskarth) [inline tests/v2/test_waku_pagination.nim:11]:
```suggestion
proc CreateSampleList(s : int) : seq[IndexedWakuMessage] =
```

File: waku/protocol/v2/waku_store.nim [hunk 2]
```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
   result.write(2, rpc.query.encode())
   result.write(3, rpc.response.encode())
 
+proc findIndex*(msgList: seq[IndexedWakuMessage], index: Index): int=
+  ## returns the position of an IndexedWakuMessage in msgList whose index value matches the given index
+  ## returns -1 if no match is found
```
Comment (oskarth) [inline waku/protocol/v2/waku_store.nim:178]: Could probably use Option type here instead of returning -1

File: waku/protocol/v2/waku_store.nim [hunk 3]
```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
   result.write(2, rpc.query.encode())
   result.write(3, rpc.response.encode())
 
+proc findIndex*(msgList: seq[IndexedWakuMessage], index: Index): int=
```
Comment (oskarth) [inline waku/protocol/v2/waku_store.nim:176]:
```suggestion
proc findIndex*(msgList: seq[IndexedWakuMessage], index: Index): int =
```

File: waku/protocol/v2/waku_store.nim [hunk 4]
```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
...
+  ## returns the position of an IndexedWakuMessage in msgList whose index value matches the given index
+  ## returns -1 if no match is found
+  for i, indexedWakuMessage in msgList:
+    if indexedWakuMessage.index  == index :
+      return i
+  return -1
+
+proc paginateWithIndex*(list: seq[IndexedWakuMessage], pinfo: PagingInfo): (seq[IndexedWakuMessage], PagingInfo) =
+  ## takes list, and perfomrs paging based on pinfo
```
Comment (oskarth) [inline waku/protocol/v2/waku_store.nim:185]:
```suggestion
  ## takes list, and performs paging based on pinfo 
```

File: waku/protocol/v2/waku_store.nim [hunk 5]
```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
...
+      newCursor = msgList[s].index # the new cursor points to the begining of the page
+
+  # retrieve the messages
+  for i in s..e:
+    result[0].add(msgList[i])
+
+  result[1] = PagingInfo(pageSize : uint64(retrievedPageSize), cursor : newCursor, direction : pinfo.direction)
+
+
+proc paginateWithoutIndex( list: seq[IndexedWakuMessage], pinfo: PagingInfo): (seq[WakuMessage], PagingInfo) =
```
Comment (oskarth) [inline waku/protocol/v2/waku_store.nim:241]:
```suggestion
proc paginateWithoutIndex(list: seq[IndexedWakuMessage], pinfo: PagingInfo): (seq[WakuMessage], PagingInfo) =
```

File: waku/protocol/v2/waku_store.nim [hunk 6]
```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
...
+  ## takes list, and perfomrs paging based on pinfo 
+  ## returns the page i.e, a sequence of WakuMessage and the new paging info to be used for the next paging request  
+  var (indexedData, updatedPagingInfo)= paginateWithIndex(list,pinfo)
```
Comment (oskarth) [inline waku/protocol/v2/waku_store.nim:244]:
```suggestion
  var (indexedData, updatedPagingInfo) = paginateWithIndex(list,pinfo)
```

File: waku/protocol/v2/waku_store.nim [hunk 7]
```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
...
+  for indexedMsg in indexedData:
+    result[0].add(indexedMsg.msg)
+  result[1]=updatedPagingInfo
+
 proc findMessages(w: WakuStore, query: HistoryQuery): HistoryResponse =
   result = HistoryResponse(messages: newSeq[WakuMessage]())
-  for msg in w.messages:
-    if msg.contentTopic in query.topics:
-      result.messages.insert(msg)
+  var data: seq[IndexedWakuMessage]= @[] # data holds IndexedWakuMessage whose topics match the query
```
Comment (oskarth) [inline waku/protocol/v2/waku_store.nim:251]:
```suggestion
  var data: seq[IndexedWakuMessage] = @[] # data holds IndexedWakuMessage whose topics match the query
```

File: waku/protocol/v2/waku_store.nim [hunk 8]
```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
...
+
+proc paginateWithoutIndex( list: seq[IndexedWakuMessage], pinfo: PagingInfo): (seq[WakuMessage], PagingInfo) =
+  ## takes list, and perfomrs paging based on pinfo 
+  ## returns the page i.e, a sequence of WakuMessage and the new paging info to be used for the next paging request  
+  var (indexedData, updatedPagingInfo)= paginateWithIndex(list,pinfo)
+  for indexedMsg in indexedData:
+    result[0].add(indexedMsg.msg)
+  result[1]=updatedPagingInfo
+
 proc findMessages(w: WakuStore, query: HistoryQuery): HistoryResponse =
```
Comment (oskarth) [inline waku/protocol/v2/waku_store.nim:249]: Should this be a public function?

File: waku/node/v2/waku_types.nim [hunk 9]
```diff
@@ -187,15 +191,34 @@ proc generateRequestId*(rng: ref BrHmacDrbgContext): string =
...
-    ctx.update(msg.contentTopic.toBytes()) # converts the topic to bytes
+  ctx.update(msg.contentTopic.toBytes()) # converts the contentTopic to bytes
   ctx.update(msg.payload)
   let digest = ctx.finish() # computes the hash
   ctx.clear()
 
   result.digest = digest
   result.receivedTime = epochTime() # gets the unix timestamp
 
+proc indexComparison* (x, y: Index): int =
```
Comment (oskarth) [inline waku/node/v2/waku_types.nim:205]: Maybe this belongs on an util folder/file?

**Review by oskarth — COMMENTED** [review]
Minor comments while looking quickly

File: waku/node/v2/waku_types.nim (hunk 9 above)
Comment (staheri14) [reply waku/node/v2/waku_types.nim:205]: Do we have such folder or file in the repo (cannot find any)? I can create a util file or move it to the https://github.com/status-im/nim-waku/blob/master/waku/protocol/v2/waku_store.nim as it is the only place the indexComparison is used

File: waku/protocol/v2/waku_store.nim (hunk 8 above)
Comment (staheri14) [reply waku/protocol/v2/waku_store.nim:249]: This is not my implementation, @decanus knows better

File: waku/protocol/v2/waku_store.nim (hunk 2 above)
Comment (staheri14) [reply waku/protocol/v2/waku_store.nim]: Thanks, changed

File: tests/v2/test_waku_pagination.nim [hunk 10]
```diff
@@ -15,14 +25,200 @@ procSuite "pagination":
...
+          1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2]
+      data3: array[32, byte] = [byte 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+          1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3]
+      
+
+      index1: Index = Index(receivedTime: 1, digest: MDigest[256](data: data1))
+      index2: Index = Index(receivedTime: 1, digest: MDigest[256](data: data2))
+      index3: Index = Index(receivedTime: 2, digest: MDigest[256](data: data3))
+      
+
```
Comment (decanus) [inline tests/v2/test_waku_pagination.nim:52]:
```suggestion

```

File: tests/v2/test_waku_pagination.nim [hunk 11]
```diff
@@ -15,14 +25,200 @@ procSuite "pagination":
...
+      indexComparison(index3, index1) == 1
+
+    check:
+      indexedWakuMessageComparison(iwm1, iwm1) == 0
+      indexedWakuMessageComparison(iwm1, iwm2) == -1
+      indexedWakuMessageComparison(iwm2, iwm1) == 1
+      indexedWakuMessageComparison(iwm1, iwm3) == -1
+      indexedWakuMessageComparison(iwm3, iwm1) == 1
+    
+    var sortingList= @[iwm3,iwm1,iwm2 ]
```
Comment (decanus) [inline tests/v2/test_waku_pagination.nim:71]:
```suggestion
    var sortingList = @[iwm3, iwm1, iwm2]
```

File: tests/v2/test_waku_pagination.nim [hunk 12]
```diff
@@ -15,14 +25,200 @@ procSuite "pagination":
...
+    sortingList.sort(indexedWakuMessageComparison)
+    check: 
+      sortingList[0] ==  iwm1
+      sortingList[1] ==  iwm2
+      sortingList[2] ==  iwm3
```
Comment (decanus) [inline tests/v2/test_waku_pagination.nim:76]:
```suggestion
      sortingList[0] == iwm1
      sortingList[1] == iwm2
      sortingList[2] == iwm3
```

File: tests/v2/test_waku_pagination.nim [hunk 13]
```diff
@@ -15,14 +25,200 @@ procSuite "pagination":
...
+      
+  
+  test "Find Index test": 
+    let
+      msgList = CreateSampleList(10)
+    check:
+      msgList.findIndex( msgList[3].index).get() == 3
```
Comment (decanus) [inline tests/v2/test_waku_pagination.nim:83]:
```suggestion
      msgList.findIndex(msgList[3].index).get() == 3
```

File: tests/v2/test_waku_pagination.nim [hunk 14]
```diff
@@ -15,14 +25,200 @@ procSuite "pagination":
...
+    sortingList.sort(indexedWakuMessageComparison)
+    check: 
+      sortingList[0] ==  iwm1
+      sortingList[1] ==  iwm2
+      sortingList[2] ==  iwm3
+      
+  
+  test "Find Index test": 
+    let
+      msgList = CreateSampleList(10)
```
Comment (decanus) [inline tests/v2/test_waku_pagination.nim:81]:
```suggestion
    let msgList = CreateSampleList(10)
```

File: tests/v2/test_waku_pagination.nim [hunk 15]
```diff
@@ -15,14 +25,200 @@ procSuite "pagination":
...
+      index1 = wm1.computeIndex()
       wm2 = WakuMessage(payload: @[byte 1, 2, 3], contentTopic: ContentTopic(1))
       index2 = wm2.computeIndex()
 
     check:
       # the digests of two identical WakuMessages must be the same
       index1.digest == index2.digest
 
+  test "Index comparison, IndexedWakuMessage comparison, and Sorting tests":
+    let
```
Comment (decanus) [inline tests/v2/test_waku_pagination.nim:39]: You should be able to do this in a nicer way.

See

```nim
var xs {.noinit.}: array[32, byte]
for x in xs.mitems: x = 1
```

File: tests/v2/test_waku_pagination.nim [hunk 16]
```diff
@@ -15,14 +25,200 @@ procSuite "pagination":
...
+      data1: array[32, byte] = [byte 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+          1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
+      data2: array[32, byte] = [byte 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+          1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2]
+      data3: array[32, byte] = [byte 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+          1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3]
+      
+
```
Comment (decanus) [inline tests/v2/test_waku_pagination.nim:47]:
```suggestion
      
```

File: tests/v2/test_waku_pagination.nim [hunk 17]
```diff
@@ -15,14 +25,200 @@ procSuite "pagination":
...
+    pagingInfo = PagingInfo(pageSize: 5, cursor: computeIndex(WakuMessage(payload: @[byte 10])), direction: PagingDirection.BACKWARD)
+    (data, newPagingInfo) = paginateWithIndex(msgList, pagingInfo)
+    check:
+      data.len == 0
+      newPagingInfo.cursor == pagingInfo.cursor
+      newPagingInfo.direction == pagingInfo.direction
+      newPagingInfo.pageSize == 0
+ 
+
+
```
Comment (decanus) [inline tests/v2/test_waku_pagination.nim:224]:
```suggestion

```

File: waku/protocol/v2/waku_store.nim [hunk 18]
```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
...
   result = HistoryResponse(messages: newSeq[WakuMessage]())
-  for msg in w.messages:
-    if msg.contentTopic in query.topics:
-      result.messages.insert(msg)
+  var data: seq[IndexedWakuMessage] = @[] # data holds IndexedWakuMessage whose topics match the query
+  for indexedMsg in w.messages:
+    if indexedMsg.msg.contentTopic in query.topics:
+      data.add(indexedMsg)
```
Comment (decanus) [inline waku/protocol/v2/waku_store.nim:254]:
```suggestion
var data = w.messages.filterIt(it.msg.contentTopic in query.topics)
```

Something like this should work.

File: waku/protocol/v2/waku_store.nim [hunk 19]
```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
...
+  result[1] = PagingInfo(pageSize : uint64(retrievedPageSize), cursor : newCursor, direction : pinfo.direction)
+
+
+proc paginateWithoutIndex(list: seq[IndexedWakuMessage], pinfo: PagingInfo): (seq[WakuMessage], PagingInfo) =
+  ## takes list, and perfomrs paging based on pinfo 
+  ## returns the page i.e, a sequence of WakuMessage and the new paging info to be used for the next paging request  
+  var (indexedData, updatedPagingInfo) = paginateWithIndex(list,pinfo)
+  for indexedMsg in indexedData:
+    result[0].add(indexedMsg.msg)
+  result[1]=updatedPagingInfo
```
Comment (decanus) [inline waku/protocol/v2/waku_store.nim:247]:
```suggestion
  result[1] = updatedPagingInfo
```

File: waku/protocol/v2/waku_store.nim [hunk 20]
```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
...
+    of PagingDirection.FORWARD: # forward pagination
+      let remainingMessages= msgList.len - foundIndex - 1
+      retrievedPageSize = min( int(pageSize), MaxPageSize).min(remainingMessages)  # the number of queried messages cannot exceed the MaxPageSize and the total remaining messages i.e., msgList.len-foundIndex
+      if initQuery : foundIndex = foundIndex - 1
+      s = foundIndex + 1  # non inclusive
+      e = foundIndex + retrievedPageSize 
+      newCursor = msgList[e].index # the new cursor points to the end of the page
+    of PagingDirection.BACKWARD: # backward pagination
+      let remainingMessages=foundIndex
+      retrievedPageSize = min( int(pageSize), MaxPageSize).min(remainingMessages) # the number of queried messages cannot exceed the MaxPageSize and the total remaining messages i.e., foundIndex-0
```
Comment (decanus) [inline waku/protocol/v2/waku_store.nim:228]:
```suggestion
      retrievedPageSize = min(int(pageSize), MaxPageSize).min(remainingMessages) # the number of queried messages cannot exceed the MaxPageSize and the total remaining messages i.e., foundIndex-0
```

File: waku/protocol/v2/waku_store.nim [hunk 21]
```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
...
+  var foundIndexOption = msgList.findIndex(cursor) 
+  if foundIndexOption.isNone : # the cursor is not valid
+    return (@[], PagingInfo(pageSize: 0, cursor:pinfo.cursor, direction: pinfo.direction))
+  var foundIndex = foundIndexOption.get()
+  var retrievedPageSize, s, e: int
+  var newCursor: Index # to be returned as part of the new paging info
+  case dir
+    of PagingDirection.FORWARD: # forward pagination
+      let remainingMessages= msgList.len - foundIndex - 1
+      retrievedPageSize = min( int(pageSize), MaxPageSize).min(remainingMessages)  # the number of queried messages cannot exceed the MaxPageSize and the total remaining messages i.e., msgList.len-foundIndex
```
Comment (decanus) [inline waku/protocol/v2/waku_store.nim:221]:
```suggestion
      retrievedPageSize = min(int(pageSize), MaxPageSize).min(remainingMessages)  # the number of queried messages cannot exceed the MaxPageSize and the total remaining messages i.e., msgList.len-foundIndex
```

File: waku/protocol/v2/waku_store.nim [hunk 22]
```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
...
+  var initQuery : bool = false
+  if cursor == Index(): 
+    initQuery = true # an empty cursor means it is an intial query
+    case dir
+      of PagingDirection.FORWARD: 
+        cursor = list[0].index # perform paging from the begining of the list
+      of PagingDirection.BACKWARD: 
+        cursor = list[list.len - 1].index # perform paging from the end of the list
+      
+
```
Comment (decanus) [inline waku/protocol/v2/waku_store.nim:211]:
```suggestion

```

File: waku/protocol/v2/waku_store.nim [hunk 23]
```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
   result.write(2, rpc.query.encode())
   result.write(3, rpc.response.encode())
 
+proc findIndex*(msgList: seq[IndexedWakuMessage], index: Index): Option[int] =
+  ## returns the position of an IndexedWakuMessage in msgList whose index value matches the given index
+  ## returns none if no match is found
+  for i, indexedWakuMessage in msgList:
+    if indexedWakuMessage.index  == index :
```
Comment (decanus) [inline waku/protocol/v2/waku_store.nim:180]:
```suggestion
    if indexedWakuMessage.index == index:
```

File: tests/v2/test_waku_store.nim [hunk 24]
```diff
@@ -57,6 +58,165 @@ procSuite "Waku Store":
...
+    discard await dialSwitch.start()
+
+    var listenSwitch = newStandardSwitch(some(key))
+    discard await listenSwitch.start()
+
+    let
+      proto = WakuStore.init(dialSwitch, crypto.newRng())
+      subscription = proto.subscription()
+      
+
```
Comment (decanus) [inline tests/v2/test_waku_store.nim:194]:
```suggestion

```

**Review by decanus — CHANGES_REQUESTED** [review]

File: tests/v2/test_waku_store.nim [hunk 25]
```diff
@@ -57,6 +58,165 @@ procSuite "Waku Store":
...
```
Comment (decanus) [inline tests/v2/test_waku_store.nim:140]:
```suggestion

```

File: tests/v2/test_waku_store.nim [hunk 26]
```diff
@@ -57,6 +58,165 @@ procSuite "Waku Store":
...
+    var subscriptions = newTable[string, MessageNotificationSubscription]()
+    subscriptions["test"] = subscription
+
+    listenSwitch.mount(proto)
+
+    for wakuMsg in msgList:
+      await subscriptions.notify("foo", wakuMsg)
+
+
```
Comment (decanus) [inline tests/v2/test_waku_store.nim:151]:
```suggestion

```

File: tests/v2/test_waku_store.nim [hunk 27]
```diff
@@ -57,6 +58,165 @@ procSuite "Waku Store":
...
```
Comment (decanus) [inline tests/v2/test_waku_store.nim:205]:
```suggestion

```

**Review by decanus — CHANGES_REQUESTED** [review]

File: waku/protocol/v2/waku_store.nim [hunk 28]
```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
...
+    pageSize = pinfo.pageSize
+    dir = pinfo.direction
+
+  if pageSize == 0: # pageSize being zero indicates that no pagination is required
+    return (list, pinfo)
+
+  if list.len == 0: # no pagination is needed for an empty list
+    return (list, PagingInfo(pageSize: 0, cursor:pinfo.cursor, direction: pinfo.direction))
+
+  var msgList:seq[IndexedWakuMessage]= list # makes a copy of the list
```
Comment (decanus) [inline waku/protocol/v2/waku_store.nim:199]:
```suggestion
  var msgList = list # makes a copy of the list
```

File: waku/protocol/v2/waku_store.nim [hunk 29]
```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
...
+  msgList.sort(indexedWakuMessageComparison) # sorts msgList based on the custom comparison proc indexedWakuMessageComparison
+
+  var initQuery : bool = false
```
Comment (decanus) [inline waku/protocol/v2/waku_store.nim:202]:
```suggestion
  var initQuery = false
```

File: waku/protocol/v2/waku_store.nim [hunk 30]
```diff
@@ -173,11 +173,89 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
...
+    initQuery = true # an empty cursor means it is an intial query
+    case dir
+      of PagingDirection.FORWARD: 
+        cursor = list[0].index # perform paging from the begining of the list
+      of PagingDirection.BACKWARD: 
+        cursor = list[list.len - 1].index # perform paging from the end of the list
+      
+  
+  var foundIndexOption = msgList.findIndex(cursor) 
+  if foundIndexOption.isNone : # the cursor is not valid
```
Comment (decanus) [inline waku/protocol/v2/waku_store.nim:213]:
```suggestion
  if foundIndexOption.isNone: # the cursor is not valid
```

**Review by decanus — CHANGES_REQUESTED** [review]

File: waku/node/v2/waku_types.nim [hunk 31]
```diff
@@ -189,15 +192,34 @@ proc generateRequestId*(rng: ref BrHmacDrbgContext): string =
...
+  ## compares x and y
+  ## returns 0 if they are equal 
+  ## returns -1 if x < y
+  ## returns 1 if x > y
+  let 
+    timecmp = system.cmp(x.receivedTime, y.receivedTime)
+    digestcm= system.cmp(x.digest.data, y.digest.data)
```
Comment (decanus) [inline waku/node/v2/waku_types.nim:213]:
```suggestion
    digestcm = system.cmp(x.digest.data, y.digest.data)
```

File: waku/node/v2/waku_types.nim [hunk 32]
```diff
@@ -189,15 +192,34 @@ proc generateRequestId*(rng: ref BrHmacDrbgContext): string =
...
+    return timecmp
+  return digestcm
+
+
+proc indexedWakuMessageComparison* (x, y: IndexedWakuMessage): int =
+  ## compares x and y
+  ## returns 0 if they are equal 
+  ## returns -1 if x < y
+  ## returns 1 if x > y
+  result= indexComparison(x.index, y.index)
```
Comment (decanus) [inline waku/node/v2/waku_types.nim:224]:
```suggestion
  result = indexComparison(x.index, y.index)
```

File: waku/node/v2/waku_types.nim [hunk 33]
```diff
@@ -189,15 +192,34 @@ proc generateRequestId*(rng: ref BrHmacDrbgContext): string =
...
+  ## returns 1 if x > y
+  let 
+    timecmp = system.cmp(x.receivedTime, y.receivedTime)
+    digestcm= system.cmp(x.digest.data, y.digest.data)
+  if timecmp != 0: # timestamp has a higher priority for comparison
+    return timecmp
+  return digestcm
+
+
+proc indexedWakuMessageComparison* (x, y: IndexedWakuMessage): int =
```
Comment (decanus) [inline waku/node/v2/waku_types.nim:219]:
```suggestion
proc indexedWakuMessageComparison*(x, y: IndexedWakuMessage): int =
```

**Review by decanus — CHANGES_REQUESTED** [review]

File: tests/v2/test_waku_pagination.nim [hunk 34]
```diff
@@ -1,11 +1,21 @@
...
-  std/unittest,
+  std/[unittest,algorithm,options],
+  nimcrypto/sha2,
+  stew/byteutils,
   ../../waku/node/v2/waku_types,
+  ../../waku/protocol/v2/waku_store,
   ../test_helpers
 
+
+proc CreateSampleList(s : int) : seq[IndexedWakuMessage] =
```
Comment (decanus) [inline tests/v2/test_waku_pagination.nim:11]:
```suggestion
proc CreateSampleList(s: int): seq[IndexedWakuMessage] =
```

**Review by decanus — CHANGES_REQUESTED** [review]

File: tests/v2/test_waku_pagination.nim [hunk 35]
```diff
@@ -1,11 +1,21 @@
...
-  std/unittest,
+  std/[unittest,algorithm,options],
+  nimcrypto/sha2,
+  stew/byteutils,
   ../../waku/node/v2/waku_types,
+  ../../waku/protocol/v2/waku_store,
   ../test_helpers
 
+
+proc CreateSampleList(s: int): seq[IndexedWakuMessage] =
```
Comment (oskarth) [inline tests/v2/test_waku_pagination.nim:11]:
```suggestion
proc createSampleList(s: int): seq[IndexedWakuMessage] =
```

File: waku/node/v2/waku_types.nim [hunk 36]
```diff
@@ -145,8 +145,11 @@ type
     #multiaddrStrings*: seq[string]
 
   WakuResult*[T] = Result[T, cstring]
+# constants required for pagination -------------------------------------------
+const MaxPageSize* = 100 # Maximum number of waku messages in each page 
+const DefaultPageSize* = 10 # The default number of waku messages in each page
```
Comment (oskarth) [inline waku/node/v2/waku_types.nim:150]: Usually constants are at the top of the file

File: waku/node/v2/waku_types.nim (hunk 9 above)
Comment (oskarth) [reply waku/node/v2/waku_types.nim:206]: Moving it to waku_store seems reasonable to me. No util folder yet, can be created if we think it will be used from multiple places.

File: waku/protocol/v2/waku_store.nim [hunk 37]
```diff
@@ -173,11 +173,84 @@ proc encode*(rpc: HistoryRPC): ProtoBuffer =
...
+  var foundIndexOption = msgList.findIndex(cursor) 
+  if foundIndexOption.isNone: # the cursor is not valid
+    return (@[], PagingInfo(pageSize: 0, cursor:pinfo.cursor, direction: pinfo.direction))
+  var foundIndex = foundIndexOption.get()
+  var retrievedPageSize, s, e: int
+  var newCursor: Index # to be returned as part of the new paging info
+  case dir
+    of PagingDirection.FORWARD: # forward pagination
+      let remainingMessages= msgList.len - foundIndex - 1
+      retrievedPageSize = min(int(pageSize), MaxPageSize).min(remainingMessages)  # the number of queried messages cannot exceed the MaxPageSize and the total remaining messages i.e., msgList.len-foundIndex
```
Comment (oskarth) [inline waku/protocol/v2/waku_store.nim:219]: Style thing but for long lines (80 chars) I'd put comment above the line

**Review by oskarth — APPROVED** [review]
Minor comments, generally lgtm!

File: tests/v2/test_waku_pagination.nim [hunk 38]
```diff
@@ -15,14 +25,193 @@ procSuite "pagination":
...
+    for x in data1.mitems: x = 1
+    var data2 {.noinit.}: array[32, byte]
+    for x in data2.mitems: x = 2
+    var data3 {.noinit.}: array[32, byte]
+    for x in data3.mitems: x = 3
+      
+    let
+      index1: Index = Index(receivedTime: 1, digest: MDigest[256](data: data1))
+      index2: Index = Index(receivedTime: 1, digest: MDigest[256](data: data2))
+      index3: Index = Index(receivedTime: 2, digest: MDigest[256](data: data3))
```
Comment (decanus) [inline tests/v2/test_waku_pagination.nim:49]:
```suggestion
      index1 = Index(receivedTime: 1, digest: MDigest[256](data: data1))
      index2 = Index(receivedTime: 1, digest: MDigest[256](data: data2))
      index3 = Index(receivedTime: 2, digest: MDigest[256](data: data3))
```

File: tests/v2/test_waku_store.nim [hunk 39]
```diff
@@ -57,6 +58,157 @@ procSuite "Waku Store":
...
+      msgList = @[WakuMessage(payload: @[byte 0], contentTopic: ContentTopic(2)),
+        WakuMessage(payload: @[byte 1],contentTopic: ContentTopic(1)),
+        WakuMessage(payload: @[byte 2],contentTopic: ContentTopic(1)),
+        WakuMessage(payload: @[byte 3],contentTopic: ContentTopic(1)),
+        WakuMessage(payload: @[byte 4],contentTopic: ContentTopic(1)),
+        WakuMessage(payload: @[byte 5],contentTopic: ContentTopic(1)),
+        WakuMessage(payload: @[byte 6],contentTopic: ContentTopic(1)),
+        WakuMessage(payload: @[byte 7],contentTopic: ContentTopic(1)),
+        WakuMessage(payload: @[byte 8],contentTopic: ContentTopic(1)), 
+        WakuMessage(payload: @[byte 9],contentTopic: ContentTopic(2))]
```
Comment (decanus) [inline tests/v2/test_waku_store.nim:178]:
```suggestion
        WakuMessage(payload: @[byte 1], contentTopic: ContentTopic(1)),
        WakuMessage(payload: @[byte 2], contentTopic: ContentTopic(1)),
        WakuMessage(payload: @[byte 3], contentTopic: ContentTopic(1)),
        WakuMessage(payload: @[byte 4], contentTopic: ContentTopic(1)),
        WakuMessage(payload: @[byte 5], contentTopic: ContentTopic(1)),
        WakuMessage(payload: @[byte 6], contentTopic: ContentTopic(1)),
        WakuMessage(payload: @[byte 7], contentTopic: ContentTopic(1)),
        WakuMessage(payload: @[byte 8], contentTopic: ContentTopic(1)), 
        WakuMessage(payload: @[byte 9], contentTopic: ContentTopic(2))]
```

**Review by decanus — CHANGES_REQUESTED** [review]

File: tests/v2/test_waku_pagination.nim (hunk 38 above)
Comment (staheri14) [reply tests/v2/test_waku_pagination.nim:49]: While the type inference absolutely works here and indeed is nicer, but curious to know whether there might be any compilation/performance issue when the explicit declaration is used?

File: waku/node/v2/waku_types.nim (hunk 9 above)
Comment (staheri14) [reply waku/node/v2/waku_types.nim:206]: At this point, it is going to be used only for pagination and in the store protocol, cannot think of a situation that it may be required for some other purposes. So, I'll move it to the waku_store

**Comment by staheri14** [comment]
Please let me know if further updates are needed on this PR, Thanks!

**Review by decanus — APPROVED** [review]
LGTM
